mesonbuild/utils/universal.py

```diff
@@ -756,40 +757,54 @@ def windows_detect_native_arch() -> str:
             raise EnvironmentException('Unable to detect native OS architecture')
     return arch
 
-def detect_vcs(source_dir: T.Union[str, Path]) -> T.Optional[T.Dict[str, str]]:
+@dataclasses.dataclass
+class VcsData:
+    name: str
+    cmd: str
+    repo_dir: str
+    get_rev: 'T.List[str]'
+    rev_regex: str
+    dep: str
+    wc_dir: 'T.Optional[str]'
+
+def detect_vcs(source_dir: T.Union[str, Path]) -> 'VcsData':
```

Comment:
This still needs to be `Optional[VcsData]`, since `None` is still returned if there is no matching vcs